flink-core/src/main/java/org/apache/flink/core/security/FileSystemWatchCertificateReloadService.java

```
@@ -0,0 +1,109 @@
+    }
+
+    @Override
+    public void run() {
+        try (WatchService watcher = FileSystems.getDefault().newWatchService()) {
+            for (String directoryPath : directoryPaths) {
+                LOG.info("Starting watching path: {}", directoryPath);
+                Path realDirectoryPath = Paths.get(directoryPath).toRealPath();
+                LOG.info("Path is resolved to real path: {}", realDirectoryPath);
+                realDirectoryPath.register(watcher, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY);
+                onWatchStarted(realDirectoryPath);
+            }
+
+            while (true) {
+                LOG.debug("Taking watch key");
+                WatchKey watchKey = watcher.take();
+                LOG.debug("Watch key arrived");
+                for (WatchEvent<?> watchEvent : watchKey.pollEvents()) {
+                    if (watchEvent.kind() == OVERFLOW) {
+                        LOG.error("Filesystem events may have been lost or discarded");
+                        Thread.yield();
+                    } else if (watchEvent.kind() == ENTRY_CREATE) {
+                        onFileOrDirectoryCreated((Path) watchEvent.context());
+                    } else if (watchEvent.kind() == ENTRY_DELETE) {
+                        onFileOrDirectoryDeleted((Path) watchEvent.context());
+                    } else if (watchEvent.kind() == ENTRY_MODIFY) {
+                        onFileOrDirectoryModified((Path) watchEvent.context());
+                    } else {
+                        throw new IllegalStateException("Invalid event kind: " + watchEvent.kind());
+                    }
+                }
+                watchKey.reset();
+            }
+        } catch (InterruptedException e) {
+            LOG.info("Filesystem watcher interrupted");
+        } catch (Exception e) {
+            LOG.error("Filesystem watcher received exception and stopped: ", e);
+            throw new RuntimeException(e);
```

Do we want to rethrow an exception here?

---

I think we want as it would be bad to stop watching for SSL certificate